File: acorn/tools/shell.py
"""Local shell command execution with background support + sandbox."""

import asyncio
import os
import re


DANGEROUS_PATTERNS = [
    'rm -rf /', 'mkfs', '> /dev/sd', ':(){:|:&};:', 'chmod -R 777 /',
]

# Known-safe commands that don't need extra approval in auto mode
SAFE_COMMANDS = {
    'ls', 'cat', 'head', 'tail', 'wc', 'sort', 'uniq', 'grep', 'find', 'which',
    'echo', 'pwd', 'whoami', 'date', 'uname', 'env', 'printenv', 'id',
    'git', 'node', 'npm', 'npx', 'yarn', 'pnpm', 'bun', 'deno',
    'python', 'python3', 'pip', 'pip3', 'uv',
    'go', 'cargo', 'rustc', 'java', 'javac', 'mvn', 'gradle',
    'ruby', 'gem', 'php', 'composer',
    'docker', 'kubectl', 'terraform',
    'make', 'cmake', 'gcc', 'g++', 'clang',
    'curl', 'wget', 'ssh', 'scp', 'rsync',
    'tar', 'zip', 'unzip', 'gzip', 'gunzip',
    'sed', 'awk', 'cut', 'tr', 'diff', 'patch',
    'mkdir', 'touch', 'cp', 'mv', 'ln',
    'chmod', 'chown',
    'sqlite3', 'psql', 'mysql', 'redis-cli', 'mongosh',
    'ffmpeg', 'convert', 'jq', 'yq',
    'nginx', 'systemctl', 'journalctl',
    'tree', 'file', 'stat', 'du', 'df', 'free', 'top', 'htop', 'ps',
    'nvidia-smi', 'rocm-smi',
    'tsc', 'eslint', 'prettier', 'jest', 'vitest', 'pytest', 'mocha',
    'kill',  # SIGTERM is fine, kill -9 caught by permissions
    'seq', 'sleep', 'true', 'false', 'test',
    'sh', 'bash', 'zsh',
}

# Paths that commands should not reference
BLOCKED_PATHS = [
    '/etc/shadow', '/etc/passwd-', '/etc/sudoers',
    '~/.ssh/id_', '~/.ssh/authorized_keys',
    '~/.gnupg', '~/.aws/credentials', '~/.kube/config',
]

def _build_blocked_patterns():
    patterns = []
    for p in BLOCKED_PATHS:
        # Match both ~ and expanded home
        patterns.append(re.compile(re.escape(p)))
        expanded = p.replace('~', os.path.expanduser('~'))
        if expanded != p:
            patterns.append(re.compile(re.escape(expanded)))
    return patterns

BLOCKED_PATH_RE = _build_blocked_patterns()

# Commands that are likely long-running (servers, watchers, etc.)
BACKGROUND_HINTS = [
    'npm start', 'npm run dev', 'npm run serve', 'yarn start', 'yarn dev',
    'python -m http.server', 'python manage.py runserver',
    'node server', 'nodemon', 'next dev', 'vite',
    'flask run', 'uvicorn', 'gunicorn', 'cargo run',
    'docker compose up', 'docker-compose up',
    'tail -f', 'watch ',
]


def get_command_binary(command: str) -> str:
    """Extract the first binary/command from a shell command string."""
    # Handle cd, env prefix, sudo, etc.
    cmd = command.strip()
    for prefix in ('sudo ', 'env ', 'nice ', 'nohup ', 'time '):
        if cmd.startswith(prefix):
            cmd = cmd[len(prefix):]
    # Handle VAR=val prefix
    while '=' in cmd.split()[0] if cmd.split() else False:
        cmd = cmd.split(None, 1)[1] if ' ' in cmd else cmd
    binary = cmd.split()[0] if cmd.split() else ''
    # Strip path
    if '/' in binary:
        binary = binary.rsplit('/', 1)[-1]
    return binary


def check_path_safety(command: str) -> str:
    """Check if command references sensitive paths. Returns error string or empty."""
    for r in BLOCKED_PATH_RE:
        if r.search(command):
            return f'Command references sensitive path: {r.pattern}'
    return ''


def _handle_bg_command(args: str, pm) -> dict:
    """Handle /bg subcommands: list, read <id>, kill <id>."""
    if not args or args == 'list':
        procs = pm.list_all()
        if not procs:
            return {'output': 'No background processes'}
        lines = []
        for bp in procs:
            status = 'running' if bp.running else f'exited ({bp.exit_code})'
            lines.append(f'#{bp.id}  {status}  {bp.elapsed}  {bp.command[:80]}')
        return {'output': '\n'.join(lines)}

    parts = args.split(None, 1)
    subcmd = parts[0]

    if subcmd == 'kill' and len(parts) > 1:
        try:
            pid = int(parts[1])
        except ValueError:
            return {'error': f'Invalid process ID: {parts[1]}'}
        if pm.kill(pid):
            return {'output': f'Killed #{pid}'}
        return {'error': f'Process #{pid} not found or already stopped'}

    # Default: treat as process ID to read output
    try:
        pid = int(subcmd)
    except ValueError:
        return {'error': f'Usage: /bg [list | <id> | kill <id>]'}

    bp = pm.get(pid)
    if not bp:
        return {'error': f'Process #{pid} not found'}

    output_lines = list(bp.output)
    status = 'running' if bp.running else f'exited ({bp.exit_code})'
    header = f'#{bp.id} [{status}] {bp.elapsed} — {bp.command[:80]}'
    if not output_lines:
        return {'output': f'{header}\n(no output captured)'}
    body = '\n'.join(output_lines[-100:])
    if len(output_lines) > 100:
        body = f'... ({len(output_lines) - 100} earlier lines)\n{body}'
    return {'output': f'{header}\n{body}'}


async def execute(input: dict, cwd: str, process_manager=None) -> dict:
    command = input.get('command', '')
    timeout_ms = min(input.get('timeout', 120000), 600000)
    timeout = timeout_ms / 1000
    background = input.get('background', False)

    # Intercept background process commands — agent can read/list/kill bg processes
    bg_match = re.match(r'^/bg\s*(.*)$', command.strip())
    if bg_match and process_manager:
        return _handle_bg_command(bg_match.group(1).strip(), process_manager)

    for pattern in DANGEROUS_PATTERNS:
        if pattern in command:
            return {'error': f'Blocked dangerous command pattern: {pattern}'}

    # Check sensitive path access
    path_err = check_path_safety(command)
    if path_err:
        return {'error': path_err}

    # Auto-detect likely long-running commands → suggest background
    is_server_like = any(hint in command for hint in BACKGROUND_HINTS)

    # If explicitly background or detected as server-like with a process manager
    if (background or is_server_like) and process_manager:
        bp = await process_manager.launch(command, cwd)
        # Wait for early output or crash — longer wait to capture server startup
        await asyncio.sleep(3.0)
        early_output = '\n'.join(bp.output) if bp.output else '(started, no output yet)'
        if not bp.running:
            return {
                'output': early_output,
                'exitCode': bp.exit_code,
                'note': f'Process exited immediately (exit {bp.exit_code})',
            }
        return {
            'output': early_output[:4000],
            'backgrounded': True,
            'processId': bp.id,
            'note': f'Running in background as #{bp.id}. To read latest output: exec /bg {bp.id}. To kill: exec /bg kill {bp.id}. To list all: exec /bg list.',
        }

    try:
        proc = await asyncio.create_subprocess_shell(
            command, cwd=cwd,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.STDOUT,
        )
        stdout, _ = await asyncio.wait_for(proc.communicate(), timeout=timeout)
        output = stdout.decode('utf-8', errors='replace')
        if len(output) > 8000:
            mid = len(output) - 8000
            output = output[:4000] + f'\n\n[... {mid} chars truncated ...]\n\n' + output[-4000:]
        return {'output': output, 'exitCode': proc.returncode}
    except asyncio.TimeoutError:
        # On timeout, move to background instead of killing if we have a process manager
        if process_manager:
            bp = await process_manager.launch(command, cwd)
            try:
                proc.kill()
            except Exception:
                pass
            return {
                'output': f'Command timed out after {timeout}s — moved to background as #{bp.id}',
                'backgrounded': True,
                'processId': bp.id,
                'note': f'To read latest output: exec /bg {bp.id}. To kill: exec /bg kill {bp.id}.',
            }
        try:
            proc.kill()
        except Exception:
            pass
        return {'error': f'Command timed out after {timeout}s', 'exitCode': -1}
    except Exception as e:
        return {'error': str(e)}

File: go/internal/tools/shell.go
package tools

import (
	"bufio"
	"context"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"runtime"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/yumlevi/acorn-cli/go/internal/bg"
)

// dangerousPatterns — exact substrings we refuse to run. Mirror
// acorn/tools/shell.py:DANGEROUS_PATTERNS.
var dangerousPatterns = []string{
	"rm -rf /",
	"mkfs",
	"> /dev/sd",
	":(){:|:&};:",
	"chmod -R 777 /",
}

// blockedPaths — sensitive file paths that any command reference is enough
// to refuse execution.
var blockedPaths = []string{
	"/etc/shadow", "/etc/passwd-", "/etc/sudoers",
	"~/.ssh/id_", "~/.ssh/authorized_keys",
	"~/.gnupg", "~/.aws/credentials", "~/.kube/config",
}

// backgroundHints — commands likely to run forever; the Python version
// auto-backgrounds these. Go port doesn't yet have a process manager, so
// we just refuse them with a useful error.
var backgroundHints = []string{
	"npm start", "npm run dev", "npm run serve", "yarn start", "yarn dev",
	"python -m http.server", "python manage.py runserver",
	"node server", "nodemon", "next dev", "vite",
	"flask run", "uvicorn", "gunicorn", "cargo run",
	"docker compose up", "docker-compose up",
	"tail -f", "watch ",
}

var blockedPathRe = buildBlockedPathsRe()

func buildBlockedPathsRe() []*regexp.Regexp {
	out := make([]*regexp.Regexp, 0, 2*len(blockedPaths))
	for _, p := range blockedPaths {
		out = append(out, regexp.MustCompile(regexp.QuoteMeta(p)))
		if strings.HasPrefix(p, "~") {
			home, _ := os.UserHomeDir()
			if home != "" {
				expanded := strings.Replace(p, "~", home, 1)
				out = append(out, regexp.MustCompile(regexp.QuoteMeta(expanded)))
			}
		}
	}
	return out
}

func checkPathSafety(cmd string) string {
	for _, r := range blockedPathRe {
		if r.FindStringIndex(cmd) != nil {
			return "Command references sensitive path: " + r.String()
		}
	}
	return ""
}

// Exec implements the exec tool. Input: command, timeout (ms), background.
// pm: optional background manager. If provided and the command is marked
// background OR matches a known-server-like pattern, launch detached and
// return a process handle instead of waiting.
// on: optional per-line output callback (used by the UI to stream exec
// output live during the call).
func Exec(input map[string]any, cwd string, logDir string, pm *bg.Manager, on func(line string)) any {
	command := asString(input["command"], "")
	if command == "" {
		return map[string]string{"error": "command is required"}
	}

	// Intercept /bg subcommands embedded in exec (agent does this to read
	// background process output via the tool layer).
	if strings.HasPrefix(strings.TrimSpace(command), "/bg") {
		return handleBgSubcommand(strings.TrimSpace(strings.TrimPrefix(strings.TrimSpace(command), "/bg")), pm)
	}

	for _, p := range dangerousPatterns {
		if strings.Contains(command, p) {
			return map[string]string{"error": "Blocked dangerous command pattern: " + p}
		}
	}
	if err := checkPathSafety(command); err != "" {
		return map[string]string{"error": err}
	}
	if pm != nil && (asBool(input["background"], false) || isServerLike(command)) {
		p, err := pm.Launch(command, cwd)
		if err != nil {
			return map[string]string{"error": "background launch failed: " + err.Error()}
		}
		// Brief wait for early-crash detection.
		time.Sleep(2 * time.Second)
		early := strings.Join(p.Output(), "\n")
		if !p.Running {
			return map[string]any{
				"output":   early,
				"exitCode": p.ExitCode,
				"note":     fmt.Sprintf("Process exited (exit %d)", p.ExitCode),
				"logFile":  p.LogPath,
			}
		}
		if len(early) > 4000 {
			early = early[:4000]
		}
		return map[string]any{
			"output":      early,
			"backgrounded": true,
			"processId":   p.ID,
			"logFile":     p.LogPath,
			"note": fmt.Sprintf("Running in background as #%d. Check output via `exec /bg %d`. Kill: `exec /bg kill %d`.",
				p.ID, p.ID, p.ID),
		}
	}

	timeoutMs := asInt(input["timeout"], 120000)
	if timeoutMs > 600000 {
		timeoutMs = 600000
	}
	inactivity := time.Duration(timeoutMs) * time.Millisecond

	// Log file setup — .acorn/logs/<ts>.log.
	var logPath string
	var logW *os.File
	if logDir != "" {
		if err := os.MkdirAll(logDir, 0o755); err == nil {
			logPath = filepath.Join(logDir, fmt.Sprintf("exec-%s.log", time.Now().Format("20060102-150405")))
			if f, err := os.Create(logPath); err == nil {
				fmt.Fprintf(f, "# Command: %s\n# Time: %s\n\n", command, time.Now().Format(time.RFC3339))
				logW = f
			}
		}
	}
	defer func() {
		if logW != nil {
			_ = logW.Close()
		}
	}()

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	shell, flag := "sh", "-c"
	if runtime.GOOS == "windows" {
		shell, flag = "cmd", "/C"
	}
	cmd := exec.CommandContext(ctx, shell, flag, command)
	cmd.Dir = cwd
	stdoutPipe, err := cmd.StdoutPipe()
	if err != nil {
		return map[string]string{"error": err.Error()}
	}
	cmd.Stderr = cmd.Stdout // merge
	stdoutPipe, _ = cmd.StdoutPipe()
	cmd.Stderr = cmd.Stdout

	if err := cmd.Start(); err != nil {
		return map[string]string{"error": err.Error()}
	}

	start := time.Now()

	// Read lines with inactivity timeout.
	var (
		mu          sync.Mutex
		lines       []string
		lineCh      = make(chan string, 1024)
		done        = make(chan error, 1)
		timedOut    bool
		totalBytes  int
	)
	go func() {
		sc := bufio.NewScanner(stdoutPipe)
		sc.Buffer(make([]byte, 0, 64<<10), 4<<20)
		for sc.Scan() {
			l := sc.Text()
			lineCh <- l
		}
		close(lineCh)
	}()
	go func() { done <- cmd.Wait() }()

	timer := time.NewTimer(inactivity)
	defer timer.Stop()

collect:
	for {
		select {
		case l, ok := <-lineCh:
			if !ok {
				// stdout EOF — wait for process exit.
				break collect
			}
			mu.Lock()
			lines = append(lines, l)
			totalBytes += len(l) + 1
			mu.Unlock()
			if logW != nil {
				fmt.Fprintln(logW, l)
			}
			if on != nil {
				on(l)
			}
			if !timer.Stop() {
				select {
				case <-timer.C:
				default:
				}
			}
			timer.Reset(inactivity)
		case <-timer.C:
			timedOut = true
			_ = cmd.Process.Kill()
			break collect
		}
	}
	err = <-done
	duration := time.Since(start).Milliseconds()

	mu.Lock()
	raw := strings.Join(lines, "\n")
	mu.Unlock()

	exitCode := 0
	if err != nil {
		if ee, ok := err.(*exec.ExitError); ok {
			exitCode = ee.ExitCode()
		} else {
			exitCode = -1
		}
	}

	if logW != nil {
		fmt.Fprintf(logW, "\n# Exit: %d, Duration: %dms\n", exitCode, duration)
	}

	if timedOut {
		return map[string]any{
			"error":    fmt.Sprintf("Command timed out after %dms of inactivity", inactivity.Milliseconds()),
			"exitCode": -1,
			"logFile":  logPath,
		}
	}

	output := raw
	truncated := false
	if len(output) > 8000 {
		mid := len(output) - 8000
		output = output[:4000] + fmt.Sprintf("\n\n[... %d chars truncated ...]\n\n", mid) + output[len(output)-4000:]
		truncated = true
	}
	result := map[string]any{
		"output":   output,
		"exitCode": exitCode,
	}
	if logPath != "" {
		result["logFile"] = logPath
	}
	if truncated && logPath != "" {
		result["note"] = fmt.Sprintf("Output truncated (%d chars). Full output: %s", len(raw), logPath)
	}
	return result
}

func isServerLike(cmd string) bool {
	for _, h := range backgroundHints {
		if strings.Contains(cmd, h) {
			return true
		}
	}
	return false
}

// handleBgSubcommand mirrors shell.py:_handle_bg_command — agent can send
// `exec /bg` / `/bg list` / `/bg <id>` / `/bg kill <id>` to inspect or
// manage background processes via the normal tool channel.
func handleBgSubcommand(args string, pm *bg.Manager) any {
	if pm == nil {
		return map[string]string{"error": "Background manager not available"}
	}
	if args == "" || args == "list" {
		procs := pm.List()
		if len(procs) == 0 {
			return map[string]string{"output": "No background processes"}
		}
		var lines []string
		for _, p := range procs {
			st := "running"
			if !p.Running {
				st = fmt.Sprintf("exited (%d)", p.ExitCode)
			}
			cmd := p.Command
			if len(cmd) > 80 {
				cmd = cmd[:80]
			}
			lines = append(lines, fmt.Sprintf("#%d  %s  %s  %s", p.ID, st, p.Elapsed(), cmd))
		}
		return map[string]any{"output": strings.Join(lines, "\n")}
	}
	parts := strings.SplitN(args, " ", 2)
	sub := parts[0]
	if sub == "kill" && len(parts) > 1 {
		id, err := strconv.Atoi(strings.TrimSpace(parts[1]))
		if err != nil {
			return map[string]string{"error": "Invalid process ID: " + parts[1]}
		}
		if pm.Kill(id) {
			return map[string]any{"output": fmt.Sprintf("Killed #%d", id)}
		}
		return map[string]string{"error": fmt.Sprintf("Process #%d not found or already stopped", id)}
	}
	// Default: treat as id to read last output.
	id, err := strconv.Atoi(sub)
	if err != nil {
		return map[string]string{"error": "Invalid /bg subcommand: " + sub}
	}
	p := pm.Get(id)
	if p == nil {
		return map[string]string{"error": fmt.Sprintf("Process #%d not found", id)}
	}
	out := strings.Join(p.Output(), "\n")
	if len(out) > 8000 {
		out = out[len(out)-8000:]
	}
	st := "running"
	if !p.Running {
		st = fmt.Sprintf("exited (%d)", p.ExitCode)
	}
	return map[string]any{
		"output":   out,
		"running":  p.Running,
		"logFile":  p.LogPath,
		"status":   st,
		"elapsed":  p.Elapsed(),
	}
}
